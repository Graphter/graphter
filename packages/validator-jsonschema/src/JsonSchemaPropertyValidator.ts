import JsonSchemaValidator from "./JsonSchemaValidator";
import { JsonSchemaValidatorOptions } from "./JsonSchemaValidatorOptions";
import {
  ModelConfig,
  PropertyConfig,
  PropertyValidator,
  ValidationExecutionStage,
  ValidationResult
} from "microo-core";
import generateErrorMessage from "./generateErrorMessage";

export default class JsonSchemaPropertyValidator extends JsonSchemaValidator implements PropertyValidator{
  constructor(options: JsonSchemaValidatorOptions) {
    super(options);
  }
  execute(stage: ValidationExecutionStage, propertyConfig: PropertyConfig, modelConfig: ModelConfig, data: any): Promise<ValidationResult> {
    const propertyValue = data[propertyConfig.id];
    const valid = this._validateFn(propertyValue);
    if(valid){
      return Promise.resolve({ valid: true });
    } else {
      const result: ValidationResult = {
        valid: false,
        errorMessage: generateErrorMessage(data, this._options, this._validateFn)
      }

      if(this._validateFn.errors){
        result.dataPaths = this._validateFn.errors.map(err => err.dataPath);
      }

      if(Array.isArray(this._options.errorDisplayMode)){
        result.errorDisplayMode = this._options.errorDisplayMode;
      } else if(this._options.errorDisplayMode !== undefined){
        result.errorDisplayMode = [ this._options.errorDisplayMode ];
      }

      return Promise.resolve(result);
    }
  }
}