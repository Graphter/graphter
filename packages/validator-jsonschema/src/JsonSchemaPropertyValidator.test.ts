import { ModelConfig, ValidationErrorDisplayMode, ValidationExecutionStage } from "microo-core";
import JsonSchemaPropertyValidator from "./JsonSchemaPropertyValidator";


describe(`JsonSchemaPropertyValidator`, () => {
  let modelConfig: ModelConfig;
  beforeEach(() => {
    modelConfig = {
      id: 'some-model',
      name: 'Some Model',
      properties: [
        {
          id: 'title',
          name: 'Title',
          type: 'string'
        }
      ]
    }
  })
  it(`should return a valid response when executed with correct data`, async () => {
    const validator = new JsonSchemaPropertyValidator({
      schema: {
        minLength: 3
      }
    });
    const data = {
      title: 'A valid page title'
    }
    const result = await validator.execute(ValidationExecutionStage.CHANGE, modelConfig.properties[0], modelConfig, data);
    expect(result).not.toBeNull();
    expect(result.valid).toBe(true);
  });
  it(`should return an invalid response when executed with incorrect data`, async () => {
    const validator = new JsonSchemaPropertyValidator({
      schema: {
        minLength: 3
      },
      error: `Title must be at least 3 characters long`
    });
    const data = {
      title: 'A'
    }
    const result = await validator.execute(ValidationExecutionStage.CHANGE, modelConfig.properties[0], modelConfig, data);
    expect(result).not.toBeNull();
    expect(result.valid).toBe(false);
    expect(result.errorMessage).toBe(`Title must be at least 3 characters long`);
  });
  it(`should error when no schema is provided`, () => {
    expect(() => {
      // @ts-ignore
      new JsonSchemaPropertyValidator();
    }).toThrowErrorMatchingSnapshot();
  });
  it(`should set executeOn values correctly`, () => {
    const validator = new JsonSchemaPropertyValidator({
      schema: {
        maxLength: 5
      },
      executeOn: [
        ValidationExecutionStage.CHANGE
      ]
    });
    expect(validator.executeOn).not.toBeNull();
    expect(validator.executeOn).toHaveLength(1);
    expect(validator.executeOn[0]).toBe(ValidationExecutionStage.CHANGE);
  });
  it(`should return correct error display mode`, async () => {
    const validator = new JsonSchemaPropertyValidator({
      schema: {
        maxLength: 5
      },
      errorDisplayMode: ValidationErrorDisplayMode.MODAL
    });
    const data = {
      title: 'Hello World!'
    }
    const result = await validator.execute(ValidationExecutionStage.CLIENT_UPDATE, modelConfig.properties[0], modelConfig, data);
    expect(result.errorDisplayMode).not.toBeNull();
    expect(result.errorDisplayMode).toEqual([ ValidationErrorDisplayMode.MODAL ]);
  })
  it(`should default options correctly`, async () => {
    const validator = new JsonSchemaPropertyValidator({
      schema: {
        maxLength: 5
      }
    });
    const data = {
      title: 'Hello World!'
    }
    expect(validator.executeOn).toMatchSnapshot();
    const result = await validator.execute(ValidationExecutionStage.CLIENT_UPDATE, modelConfig.properties[0], modelConfig, data);
    expect(result.errorDisplayMode).toMatchSnapshot()
  });
});

export {};