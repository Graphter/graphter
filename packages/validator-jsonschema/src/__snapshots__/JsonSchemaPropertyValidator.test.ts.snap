// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchemaPropertyValidator should default options correctly 1`] = `
Array [
  "CLIENT_CREATE",
  "CLIENT_UPDATE",
  "SERVER_UPDATE",
  "SERVER_UPDATE",
]
`;

exports[`JsonSchemaPropertyValidator should default options correctly 2`] = `
Array [
  "INLINE",
  "SUMMARY",
]
`;

exports[`JsonSchemaPropertyValidator should error when no schema is provided 1`] = `"Schema option is required by JsonSchemaValidator"`;
